content/05.dl-complexities.md

```diff
@@ -1 +1,14 @@
 ## Tip 3: Understand the complexities of training deep neural networks {#complexities}
+
+Correctly training deep neural networks is a non-trivial process.
+There are many different options and potential pitfalls at every stage.
+To get good results you have to expect to train many networks with different tweaks and modifications.
+Despite improved framework ease-of-use and on-demand cloud computing resources this means DL can be very demanding, requiring significant infrastructure and patience to complete.
+The experimentation inherent to DL is often noisy (requiring repetition) and represents a significant organizational challenge.
+All code, parameters, and results must be carefully corralled using good coding practices (for example, version control, continuous integration etc.) in order to be effective and interpretable.
```

Comment:
I'm curious about examples of continuous integration with DL.  Is there anything we could refer to?  Given that model training and performance can depend on things like the hardware and versions of required packages, I am interested in best practices here and think our readers would be as well.

Reply:
We definitely to mention non-deterministic algorithm choices in CUDA & cuDNN. I.e., by default in e.g., TensorFlow and PyTorch, running code on the GPU (as opposed to CPU) is non-deterministic usually (esp. convolutions). This is because cuDNN has many different approximations of convolutions implemented (i.e., all the winnograd stuff and derivates thereof approximating convolutions via fast fourier transforms). So, by default, if you run GPU code, the "best" algorithm is chosen at runtime and may vary from run to run (and can also change during runtime I think). Hence, it's best practice for reproducibility to set `torch.backends.cudnn.deterministic = True` in PyTorch, for example (forgot how to do that in TensorFlow). In practice, it's also often faster as the "guessing" by cuDNN is not that good (heard that from someone at NVIDIA as well when I reported it). 

Note that setting random seeds will only affect things like deterministic shuffling and weight initialization, the algorithm determinism has to be set on top of that. I think this is important to mention. We could reference the cuDNN technical manuals for this: https://docs.nvidia.com/deeplearning/sdk/cudnn-developer-guide/index.html#reproducibility

Reply:
Couldn't really find a good CI example specifically for DL usage instead of writing new modules/frameworks.   

Good call on the non-deterministic/deterministic backend algorithms part, I'd forgotten totally about that.